noir-projects/aztec-nr/aztec/src/encrypted_logs/encrypted_note_emission.nr

```diff
@@ -123,6 +101,9 @@ pub fn encode_and_encrypt_note_with_keys_unconstrained<Note, let N: u32, let NB:
     recipient: AztecAddress
 ) -> fn[(&mut PrivateContext, OvpkM, IvpkM, AztecAddress)](NoteEmission<Note>) -> () where Note: NoteInterface<N, NB>, [Field; N]: LensForEncryptedLog<N, M> {
     | e: NoteEmission<Note> | {
-        emit_with_keys(context, e.note, ovpk, ivpk, recipient, compute_unconstrained);
+        let (note_hash_counter, encrypted_log, log_hash) = unsafe {
+            compute_raw_note_log_unconstrained(*context, e.note, ovpk, ivpk, recipient)
+        };
+        context.emit_raw_note_log(note_hash_counter, encrypted_log, log_hash);
```

Comment:
This seems odd. `encrypted_log` is definitely unconstrained, but `note_hash_counter` and `log_hash` do not need to be unconstrained. I'm not sure what implications being able to pass arbitrary values there would be however - could an attacker manipulate the logs in some unexpected way?

If not, then we need to have a long comment here explaining why it is ok to call this unconstrained function (within this context, namely that we're assuming a cooperative sender that'll encrypt properly, and that these values cannot be used to break stuff). If it is not ok to receive an unconstrained `note_hash_counter` or `log_hash`, then we need a small refactor here.

Reply:
Very good catch. Not sure how I missed this. The unsafe blocks are indeed very helpful here.

The log hash was unconstrained in [the original code as well](https://github.com/AztecProtocol/aztec-packages/blob/365193617179fbbfb6a19c8f194a0a214beac87f/noir-projects/aztec-nr/aztec/src/encrypted_logs/encrypted_note_emission.nr#L49). Will investigate.

Reply:
It seems like having the log hash be unconstrained in the case of unconstrained log encryption really is fine because if the hash was non-sense the tx would be dropped by the sequencer (the sequencer apparently hashes the note encrypted logs somewhere out of circuit and checks that it matches the log hash).

So essentially the sender here is trusting his PXE that it computes the log hash correctly and if it didn't the tx would be dropped. This seems like a much better outcome for the sender then in the case of his PXE computing invalid encrypted log with valid log hash of the invalid encrypted log because in the case of invalid encrypted log only the tx would get included and the other party receiving could decline considering the payment as valid.

This makes me realize that the unconstrained encryption flow might be problematic because unless we implement some kind of fraud proof scheme the sender is making himself vulnerable against the recipient claiming that the encrypted log is invalid even when valid. Let's consider the payment flow. I send a token transfer which does unconstrained encryption. I am well behaved and I computed the encrypted log correctly.

Now the recipient is malicious and says that the log is invalid. In the current fiat world in such a situation I would take a bank receipt and I could use that in the legal system to prove that I did the payment.

In the zk world I would have to show to some legal entity that I really sent the tx by decrypting the tx info for them and them verifying it. Given that they would have to learn about a private info here to be able to resolve the issue they could also just re-encrypt the note and see that the log was legit.

So there really seems to be no problem with the log hash being invalid here.

Will confirm with Leila if it's the case with the counter as well and if yes will put there the comment.

Reply:
Just spoke to Leila and confirmed that this is fine. It should be noted however that we should not be thinking about cooperative senders and malicious recipients but rather malicious senders, since they are the ones making the proof.

The way this works is we send the log hash to the kernel, and it gets included as part of its public inputs. Then we send the tx to the sequencer, which includes the kernel proof and the log preimages. The sequencer computes the hashes of the logs and checks that they are the ones in the public inputs of the kernel, and drops the tx otherwise (circuits will later fail to solve if it doesn't).

So if we don't constrain the log hash, then a malicious sender can compute the correct log, submit a bad log hash to the kernel, and then submit the bad log preimage to the sequencer. All checks will pass, but the submitted log will *not* be the one that was computed by the app.

In the unconstrained case, we don't care about the log at all because we don't do anything with it, and because it's unconstrained: it could be anything. So if a sender chooses to broadcast the tx with a log that is different from the one that was used in the circuit, then they'll be able to, but they were already able to change the log before anyway, so the end result is the same. It's important here that we do not return the log from this function to the app, otherwise it could try to do stuff with it and then that might be wrong.

Re. the note hash counter, this is used for squashing. The kernel assumes that a given note can have more than one log and removes all of the matching ones, so all a malicious sender could do is either: cause for the log to be deleted when it shouldn't have (which is fine - they can already make the content be whatever), or cause for the log to not be deleted when it should have (which is also fine - it'll be a log for a note that doesn't exist).

tldr unconstrained for unconstrained logs good, unconstrained for constrained logs bad.

Reply:
Nice explanation, put the comment inside the code. I think you just did one small mistake here:

> he sequencer computes the hashes of the logs and checks that they are the ones in the public inputs of the kernel, and drops the tx otherwise **(circuits will later fail to solve if it doesn't)**.

As far as I know we don't ever feed the log preimages to any of the protocol circuits because they are variable length. Instead we hash the preimages inside our L1 contracts with [this function](https://github.com/AztecProtocol/aztec-packages/blob/e52799256d67e2f992c141784102e9b97b114e49/l1-contracts/src/core/libraries/decoders/TxsDecoder.sol#L300). Once we have the kernel logs hash we do the same thing as the rollup circuits do and we hash the tx effects to get the final `txs effects hash` for the block. Then in [Rollup contract](https://github.com/AztecProtocol/aztec-packages/blob/932dcf5cb33109a28e8dce1fbf525bd7e01d4c03/l1-contracts/src/core/Rollup.sol#L645) we check that the txs effects hash from rollup circuits matches the one computed on L1.

Reply:
Ah, I see. Well, the rollup rejecting the proof is almost the same as that :stuck_out_tongue: